tools/bitfield_gen.md

```diff
@@ -0,0 +1,653 @@
+uint32_t VMFault_ptr_get_seL4_FaultType(VMFault_t *vm_fault);
+```
+
+These are declared `PURE`, but not `CONST`, since they depend on global state.
+
+The pointer set functions are:
+
+```c
+void VMFault_set_address(VMFault_t *vm_fault, uint32_t address);
+void VMFault_set_FSR(VMFault_t *vm_fault, uint32_t FSR);
+void VMFault_set_instructionFault(VMFault_t *vm_fault, uint32_t instructionFault);
+void VMFault_set_seL4_FaultType(VMFault_t *vm_fault, uint32_t seL4_FaultType);
+```
+
+These are neither `PURE` nor `CONST`, but still declared as `static inline`.
+
+### Field High
+
+The usual access patterns for fields is that if we provide e.g. a value of 5 to
+a setter function, we expect a corresponding 8-bit field to be set to the bit
+pattern `0000 0101`.
+
+In general, on a 32-bit architecture, we are providing a 32-bit value to store
+in an n-bit field. One way of looking at this is that we know the top 32-n bits
+of the value are always 0.
+
+Sometimes we instead know that the *bottom* n bits of the value we provide are
+0. This can happen for instance when we want to store an address that is always
+aligned to a page boundary, as it would in a page table entry. This means we are
+interested in the high bits of the address. The `field_high` tag allows us to
+store only the significant high bits of such addresses.
+
+For instance the base pointer stored in a page table cap in seL4 for 32-bit
+RISC-V is such a case:
+
+```bf_gen
+block page_table_cap {
+    field       capPTMappedASID     9
+    field_high  capPTBasePtr        20
+    padding                         3
+
+    padding                         7
+    field       capPTIsMapped       1
+    field_high  capPTMappedAddress  20
+    field       capType             4
+}
+```
+
+The generated access methods of the `field_high capPTBasePtr` have the same type
+as the ones for `field`, but only the top 20 bits of the provided value will be
+stored in the setter methods as opposed to the bottom 20 bits for normal fields.
+
+The getter methods will automatically correctly return a value that has the top
+20 bits set to the stored value and the bottom 12 bits set to 0. That is, getter
+and setter remain inverses of each other as long as the values provided to the
+setter have the bottom 12 bits set to 0.
+
+On most 64-bit architectures, the top bits of pointers are also determined,
+because not all 64 bits are available (e.g. only 48 of them on x64). The
+generator assumes `field_high` values to be used for pointers and to need to store
+only the significant bits of the pointer, along with potential sign extension.
+See the [Architecture Parameters][] section below for details.
+
+For instance, if the platform is declared to have base 64 and 48 significant
+bits, and we have a pointer that we know is aligned to 12 bits, we need to store
+only 48-12 = 36 bits to recreate the correct original pointer value. That means
+if we declare
+
+```bf_gen
+block store_more {
+    field       some_info  26
```

Comment:
Was this supposed to add up to `64` total bits? in which case some_info should probably be `16`.  I don't think it breaks the example though it's just odd to be declaring a block with 74 total bits.

Reply:
Yes, totally was. Don't trust verification people with adding numbers :-)